### ISSUE TITLE
[AOB XI ] Cache Cleaning script - (housekeeping)

### ISSUE THREAD

**Issue by simboss** [geosolutions-it/eumetsat-EOWS#389]
From client requirements:

> Implement a simple software component (a script would do) to housekeep the granules in GeoServer. The software component should run independently from the other components, and it shall be driven by a simple configuration file specifying the retention period (granularity: hours) per layer. The component shall run periodically and it shall also be possible to run the component manually.

**Comment by simboss** [comment]
From the client:

> Fabrizio Borgia commented on W9-9:
> ----------------------------------
> 
> For completeness and future reference: a basic cache housekeeping script has been put in place on machine dsvv-01 (in crontab, executed every hour). The content of the script is the following.
> 
> The script will be removed once the housekeeping from GeoSolutions is in place.
> 
> 
> 
> {noformat}#!/bin/bash
> 
> #======================================================================================
> # MANUAL_CONFIGURATION
> #======================================================================================
> 
> # Log configuration.
> LOG_DIRECTORY='/tcenas/home/borgiaf/log'
> LOG_FILE="$LOG_DIRECTORY/EUMETView_housekeep_cache_by_time.$(date +%Y-%m-%d).log";
> LOG_GLOB="$LOG_DIRECTORY/EUMETView_housekeep_cache_by_time.*"
> LOG_RETENTION=3
> 
> # Namespace and name of the cache pod.
> CACHE_NS='wms'
> CACHE_POD='mapcache-0'
> CACHE_CONTAINER='mapcache'
> 
> # Name of the cache partition within the pod.
> ROOT_CACHE_DIR="/cache"
> 
> # Cache retention time (in days).
> CACHE_RETENTION="1"
> 
> #======================================================================================
> # MAIN LOGIC
> #======================================================================================
> 
> exec >> $LOG_FILE
> 
> echo "-------------------------------------------------------------------------------------"
> echo "[$(date +%Y-%m-%d" "%H:%M:%S)] Starting cache housekeeping."
> 
> # Housekeeping old logs.
> find $LOG_DIRECTORY -name $LOG_GLOB -mtime +${LOG_RETENTION} -delete.
> 
> # Load the environment to connect to the Kubernetes cluster.
> export KUBECONFIG=~/.kube/icsi-olda-ope
> 
> # Housekeeping cache.
> kubectl -n ${CACHE_NS} exec ${CACHE_POD} -c ${CACHE_CONTAINER} -- find $ROOT_CACHE_DIR -maxdepth 1 -mindepth 1 -not -name "*lost*" -not -name "*now*" -type d -mtime +${CACHE_RETENTION} -exec echo "[$(date +%Y-%m-%d" "%H:%M:%S)] Housekeeping {}." \; -exec rm -rf {} \;
> 
> echo ""{noformat}

**Comment by randomorder** [comment]
Hey @jemacchi 

Unlike what Fabrizio Borgia did in his script above, we don't want to use the modification time of the files to decide wether or not to remove the file, we want to parse the TIMEstamp that is embedded in the name of the directories:
![image](https://user-images.githubusercontent.com/5264230/104470203-c194d780-55b9-11eb-84ca-88e9e8b6b45d.png)

I would use a [CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) for the purpose, assuming it's available for the K8s version they are running in the various environments (DEV, VAL and OPE. Note that they are running 3 diff. versions)

You don't necessarily have to stick to Bash, you can use Python if you prefer (timestamp parsing can be tricky). Just make sure that it is reasonably performant and configurable because there's potentially a lot of tiles to remove.

**Comment by randomorder** [comment]
Adding to the above @jemacchi 
We want to have visibility on what the script is doing in terms of logging

This script will take care of the Mapcache cached tiles. For GWC cached tiles I'll create a separate issue

**Comment by jemacchi** [comment]
@randomorder 
Status: I have all the main bash commands which allows to handle date parsing ready, so now working on a script that "merges" that and allow me to clean caches. Hope today to get the main script done, and then testing as much as possible that in dev environment (ETA: maybe today/tomorrow and I will get it done, if I dont find any trouble)

script configuration file will look like (named ie. layers.properties):
```
...
layername=48
layername2=24
...
```

So script will check each layer configuration, and apply a clean based on the retention time (in hours) specified in that properties file. (ie. in the sample, will check for all the cache folders containing tiles for layer layername and it will remove those folders based on date(NOW) - retentiontime*3600 )

**Comment by jemacchi** [comment]
@randomorder 
[Here](https://github.com/geosolutions-it/eumetsat-EOWS/pull/463) you can find the PR about script.

I think that it covers what we expect from it. Tested on DEV environment in mapcache pods and it's working (manual execution)
I will need your assistance regarding setting it from kubernetes.

**Comment by randomorder** [comment]
What about the [kubernetes CronJob](https://github.com/geosolutions-it/eumetsat-EOWS/issues/389#issuecomment-759514145) @jemacchi ?

**Comment by jemacchi** [comment]
> What about the [kubernetes CronJob](https://github.com/geosolutions-it/eumetsat-EOWS/issues/389#issuecomment-759514145) @jemacchi ?

That's something I have to read about on how it should be implemented in K8S (that is reason I mentioned [here](https://github.com/geosolutions-it/eumetsat-EOWS/issues/389#issuecomment-763108235) that probably I will need your support, just to be pointed in right direction, and avoid spending more effort than necessary)

Correct me if wrong, but I should use something like this https://kubernetes.io/docs/tasks/job/automated-tasks-with-cron-jobs/ , right ? (I have the script and config file, need to check where to place them and allow k8s to call them)

**Comment by randomorder** [comment]
You should have the information you need to move on now. Keep us posted

**Comment by simboss** [comment]
@jemacchi we need to speed this up, it should have been completed by now.

**Comment by jemacchi** [comment]
@randomorder here the status:

Locally I have build an docker image that contains script and pushed it to Harbor.

```
docker build -t harbor.cgi-dev.de/geoserver/mapcache-cronjob:1.0 .
docker run --rm -it -v /home/jmacchi/cache:/cache harbor.cgi-dev.de/geoserver/mapcache-cronjob:1.0 /cache
docker push harbor.cgi-dev.de/geoserver/mapcache-cronjob:1.0
```

In the jumpbox, go to cronjbs-test folder and you will find all the required files (you will be interested mainly in the yaml)
With this command you can create the cronjob

`kubectl create -f mapcache-cronjob.yaml
`
then you can check jobs and related pods

```
kubectl get jobs -n wms
kubectl get pods -n wms
```

Finally you can remove the cronjob with this command
`kubectl delete -n wms cronjob mapcache-cleancache`

Since our new container **(mapcache-cleancache)** should access the same pvc as the mapcache container (in order to access the /cache folder where tiles are saved), then tried to edit the pvc to change the ReadWriteOnce for a ReadWriteMany (found that it was required while I was trying to test in dev env), but it's not possible since the status of such kind of elements are immutable once defined (showing this error "spec: Forbidden: is immutable after creation except resources.requests for bound claims")

Edit command below:
`kubectl -n wms edit pvc mapcache-tiles-wms-geoserver-mapcache-0`

I think that once that pvc is accesible by both containers (mapcache one writing tiles and the clean-cache one reading and removing files)  we will be able to see the logs for those pods created for the cronjob.

In order to check if it's working, in another shell, you can go inside mapcache POD 

`kubectl exec -it -n wms wms-geoserver-mapcache-0 -- bash
`
and execute
`watch find /cache -type d -name msg_fes:rgb_airmass `

(if you navigate the Eumetsat client in a browser with different times, then you will see how the list of items increases based on TIME dimension)

I tested scripts manually inside of the mapcache container, and also running in the dockerized cronjob container. Here what it's needed it's to finally set it on the development environment using kubernetes (sorry  but I'm not an expert on it, and it's really  making testing a little more complex).
**In general, Yaml, docker and scripts files are done**. Code will be at helm-charts repo, branch cache-clean-script-cronjob, but I was going to push it to repo but I found that also the repo https://gitlab.cgi-dev.de/eows/wms/helm-charts.git/ is down after 18pm my time.

One last thing (that is pending, since I did not know how to do it and probably it implies a new yaml file): I believe that as much as the cronjob container should access the cache tiles folder, we need to add a configmap or something that would allow to map the  layers.properties file outside of the container (that way retention times can be updated "easily" compared with recreating the docker image). Correct me if I'm wrong.

**Comment by randomorder** [comment]
> Since our new container **(mapcache-cleancache)** should access the same pvc as the mapcache container (in order to access the /cache folder where tiles are saved), then tried to edit the pvc to change the ReadWriteOnce for a ReadWriteMany (found that it was required while I was trying to test in dev env), but it's not possible since the status of such kind of elements are immutable once defined (showing this error "spec: Forbidden: is immutable after creation except resources.requests for bound claims")

This is a problem. I don't think we I'll be able to change the access mode without changing the storage class of the volume and impact the performance of mapcache and that would be a no-no. Let's hear from the K8 administrators (CGI)

**Comment by jemacchi** [comment]
> Let's hear from the K8 administrators (CGI)

@randomorder  do I have to ping them ? (if so, let me know who is in charge and I can ask).

**Comment by jemacchi** [comment]
> Let's hear from the K8 administrators (CGI)

@randomorder  I see that you asked to Nicolas and Kiril (email subject:  Access Cache Volumes from Cronjob )
Waiting for their answer.

**Comment by jemacchi** [comment]
@simboss  @randomorder 
Based on email thread with K8 admins we moved to a different approach using sidecar containers (instead of using K8 cronjobs)

**It's working fine in development environment.**

Work included:
- create a new docker image, which includes:
    - necessary setup to run cronjobs (defined as parameters in docker containers, so we can set cronjobs externally)
    - cleaning mapcache script that removes tiles based on mapcache folder structure (parsing time dimension)
    - layers.properties (that defines layers=retention time kvps, used by the script. Note: this file is overriden externally by kubernetes, using a configmap)

- created a new cm called mapcache-cronjob-layers 

`kubectl create configmap mapcache-cronjob-layers --from-file layers.properties -n wms`

- edited/configured the statefulset for mapcache adding a new docker container to the pod (based on developed image) which executes a cronjob every minute checking layers to be clean based on mapcache-cronjob-layers configmap

**kubectl -n wms edit sts wms-geoserver-mapcache**
(code included in the sts)
```
...
- image: harbor.cgi-dev.de/geoserver/mapcache-cronjob:1.1
        imagePullPolicy: Always
        name: mapcache-cronjob
        args:
        - start-cron 
        - "\\* \\* \\* \\* \\* /home/cleancache.sh /cache /home/layers.properties >> /var/log/cron.log 2>&1"
        volumeMounts:
        - mountPath: /cache
          name: mapcache-tiles
        - mountPath: /home/layers.properties
          name: mapcache-cronjob-layers
          subPath: layers.properties
... 
   volumes:
      - configMap:
          defaultMode: 450
          name: mapcache-configmap
        name: mapcache-configmap
      - configMap:
          defaultMode: 450
          name: mapcache-cronjob-layers
        name: mapcache-cronjob-layers
...
```

Work to do: configure list of all the layers and retention times (you can check current list running following command in jumpbox: 

` kubectl describe configmap mapcache-cronjob-layers -n wms`

**Comment by randomorder** [comment]
Hi @jemacchi 

>        - "\\* \\* \\* \\* \\* /home/cleancache.sh /cache /home/layers.properties >> /var/log/cron.log 2>&1"

- What happens if `cleancache.sh` takes longer than a minute? Are we going to have multiple `cleancache.sh` running in parallel and piling up?
- Is `/var/log/cron.log` ever rotated?
- Can we have instead the script log to stdout / stderr so it can be monitored with `kubectl logs ...`?


I had a look at `cleancache.sh` and it look ineficient to me. Remember we're dealing with **MANY** files and directories here, I see:
```
input=${2}

HOUR=3600 #secs in 1 hour
now_epoch=$(date +%s)
echo "----------------------------------------------------------------------"
date -d @$now_epoch | xargs echo "Starting cleaning process with time = "
echo "Cache root folder = $1"

path_slashes=$(echo "${1}" | awk -F"/" '{print NF-1}')
let path_depth=$path_slashes+2

echo "..."
if [ -f "$input" ]
then
  echo "Layers retention time properties file: $input found."
  # outer while: reading each layer retention time setting (granularity: hours)
  while IFS='=' read -r key value
  do
    #eval ${key}=\${value}
    # inner while:
    echo "..."
    echo "Cleaning cache for layer: $key - Retention time: $value hours"
    let diff_epoch=$now_epoch-$value*$HOUR
    date -d @$diff_epoch | xargs echo "Checking cache folders with TIME dimension previous to "
    echo "Detected folders for $key"
    find ${1} -type d -name $key
    echo "..."
    find ${1} -type d -name $key -print0 |
    while IFS= read -r -d '' line; do
        echo "   Folder in revision: $line"
        folder_time_dimension_epoch=$(echo $line | sed -r 's/[#]+/_/g' | cut --delimiter="/" -f${path_depth} | cut --delimiter="_" -f3 | xargs -I{} date -d {} +%s)
        if [ ! -z "${folder_time_dimension_epoch}" ]
        then
           if [ $folder_time_dimension_epoch -lt $diff_epoch ]
           then
             echo "   --> Deleting folder $line"
             rm -rdf $line
           fi
        fi
    done
  done < "$input"
else
  echo "Layers retention time properties file $file not found."
fi
echo "----------------------------------------------------------------------"root@wms-geoserver-mapcache-0:/#
```

My understanding is (correct me if I am wrong, I am no bash guru):

For each line in the properties file, two `find` commands are executed **at the root** (`$1`) of the cache directory and the first "find" only purpose is to print out the directories that matched. 

You know how long this is going to take in production?
Currently (and the system is NOT heavily used) each `find` run takes about 5minutes, so:
5 x 2 (double find) x #of layers in the properties file (which ATM is ~60 is I remember correctly) = 600 minutes
At that is just the `find` command execution. The deletion of the tiles is not taken into account

**Comment by jemacchi** [comment]
@randomorder 

- cronjob can be adjusted, I'm testing with every minute running, but we can set it 1 per day (? I think that we should ask Cesare/Nicolas and let them decide)
- currently you can see the logs with kubectl logs ... 
- i can check how to rotate the logs file
- first find can be removed (it was just for the purpose of testing. No need to log that list, it was only useful for me during development)
- no idea about production conditions about folders since I dont have access to prod environment, but I dont think that an approach using - as an example - python would be better than bash.  
- not sure how it would be possible to improve performance, since basically structure is based first on TIME dimension and then in layers names, so we need -for each layer- to get all the folders which are inside of the time dimentions folders. May be an alternative way would be to read the configuration file, group all the layers that share the same retention-time and then to cycle in the TIME folders which parsed retention time has expired (so then clean all folder layers that matches list of layers). In any of the cases, deletion time will be the same (it's the minimal atomic action, each tiles folder is deleted one by one based on time+layer name).

**Comment by randomorder** [comment]
> @randomorder

> * currently you can see the logs with kubectl logs ...
> * i can check how to rotate the logs file

Cool, please make the logging to file optional and disable it by default. No need to investigate the rotation

> * first find can be removed (it was just for the purpose of testing. No need to log that list, it was only useful for me during development)
> * no idea about production conditions about folders since I dont have access to prod environment, but I dont think that an approach using - as an example - python would be better than bash.  BTW, if I can get access to prod, I can try with some extra parameters in find command (sample, defining maxdepth parameter will allow to tail the search)

You don't need access to OPE to estimate the number of tiles nor to test the find flags but if you do I wish you had asked before

> * not sure how it would be possible to improve performance, since basically structure is based first on TIME dimension and then in layers names, so we need -for each layer- to get all the folders which are inside of the time dimentions folders. May be an alternative way would be to read the configuration file, group all the layers that share the same retention-time and then to cycle in the TIME folders which parsed retention time has expired (so then clean all folder layers that matches list of layers). In any of the cases, deletion time will be the same (it's the minimal atomic action, each tiles folder is deleted one by one based on time+layer name).

Ok but the directory structure is predictable. You can leverage that, no? Very first thing comes to my mind
(just an example, to explain my point. You can try to estimate how long this is going to take and you should come up with your own solution):
- have the <layers, retention> relationship in a data structure in memory (bash array, python tuple, whatever)
- for every time available (only 1 listing of the root of `/cache` needed) compute the list of layers that you may have to cleanup based on the retention time and do `rm -rf #TIME123/layername`

If the layer for that time does not exist `rm -rf` will not bother

**Comment by jemacchi** [comment]
@randomorder 
I have a version of the script based on assembling a list of layers for the same retention time and then to remove the folders for each time-dimension folder (that avoids to use the find command, since I'm using a for -much faster, based on comments I read- combined with rm -rf list-of-folders).  
Also I have a different version that allows to remove based on rsync --delete (since in some documentation it's commented that is much more convinient when you have tons of files).

BTW, also I have the initial approach (same script as you saw before) using a -maxdepth 2 in the find command. I have been comparing times (time command) between executions in local environment, and that reduced 3 times the execution time.(good improvement with minor change on code)

Logging to file can be disabled in the cronjob command line (no need to modify the scripts)

In order to continue, I need to run those scripts on development environment. At the end of today, DEV environment continued to be unusable (client was not working so cannot generate bunch of tiles in cache as before)

**Comment by jemacchi** [comment]
Some notes:
- I have checked find command comparing with -maxdepth 2 and without it, and it works 2-3x times faster if maxdepth param is defined.
- using `- "\\* \\* \\* \\* \\* /home/cleancache.sh /cache /home/layers.properties >  /proc/1/fd/1 2>&1"`   in the configmap for the crobjob definition allows to avoid saving logs in a file (instead, it's redirected to proc1/fd/1 and then it's possible to look at logs running  `kubectl -n wms logs wms-geoserver-mapcache-0 mapcache-cronjob`
- also, definition in configmap should be changed once moved the implementation to OPE system (that is: changing * * * * * values to make it running one time a day -?-)

**Comment by jemacchi** [comment]
@randomorder 
last 2 commits [here](https://gitlab.cgi-dev.de/eows/wms/helm-charts/commits/cache-clean-script-cronjob) contain the new scripts version as much as configuration for layers.properties (based on retention_time=48 hs, and including all layers cached by mapcache.  We need to check/confirm retention time with @simboss )

**Comment by randomorder** [comment]
@jemacchi 
for the retention please check the expire values in the mapcache config [here](https://gitlab.cgi-dev.de/eows/wms/helm-charts/blob/cache-clean-script-cronjob/charts/wms-geoserver/configFiles/mapcache.xml)

```
<!-- auto_expire

        Tiles older (in seconds) than this value will be re-requested and
        updated in the cache. Note that this will only delete tiles from the
        cache when they are accessed: You cannot use this configuration to
        limit the size of the created cache. Note that, if set, this value
        overrides the value given by <expires>. By default tiles don't expire.
   -->
```

**Comment by jemacchi** [comment]
@randomorder 
> for the retention please check the expire values in the mapcache config [here](https://gitlab.cgi-dev.de/eows/wms/helm-charts/blob/cache-clean-script-cronjob/charts/wms-geoserver/configFiles/mapcache.xml)

updated to 720 hs (it's the current value in hours at mapcache. 720hs = 30 days. Those tiles older than 30 days will be removed by the script, and invalidated at the same time by mapcache)

**Comment by jemacchi** [comment]
@randomorder created the [MR](https://gitlab.cgi-dev.de/eows/wms/helm-charts/merge_requests/23) in gitlab

**Comment by randomorder** [comment]
@jemacchi please check my feedback in the MR and keep an eye on the JIRA issue. I've added you as a watcher
https://cgi-pathfinders.atlassian.net/browse/W9-24

I want to merge it in the early afternoon

**Comment by randomorder** [comment]
Hi @jemacchi 
I am testing this in DEV. Looks like it's doing the job but I would like you to improve a couple of things:
1 - logging
2 - directories pruning / possible speedup

### For 1:
The following is extracted from a cache_clean run:
```
...
   --> Deleting layers folders from time-dimension /cache/#TIME#2020-11-02T21:00:00#000Z with retkey > 720

   --> Layers:  ./backgrounds:ne_10m_coastline/ ./backgrounds:ne_background/ ./backgrounds:ne_boundary_lines_land/ ./backgrounds:ne_gray/ ./eps:m01_ir108/ ./eps:m01_rgb_124/ ./eps:m01_rgb_natural_fog/ ./eps:m02_ir108/ ./eps:m02_rgb_124/ ./eps:m02_rgb_natural_fog/ ./eps:m03_ir108/ ./test:l3/ ./msg_fes:clai/ ./msg_fes:clm/ ./msg_fes:cth/ ./msg_fes:fire/ ./msg_fes:h03b/ ./msg_fes:ir039/ ./msg_fes:ir108/ ./msg_fes:rgb_airmass/ ./msg_fes:rgb_ash/ ./msg_fes:rgb_convection/ ./msg_fes:rgb_dust/ ./msg_fes:rgb_eview/ ./msg_fes:rgb_fog/ ./msg_fes:rgb_microphysics/ ./msg_fes:rgb_natural/ ./msg_fes:rgb_naturalenhncd/ ./msg_fes:rgb_snow/ ./msg_fes:rgb_tropicalairmass/ ./msg_fes:vis006/ ./msg_fes:wv062/ ./msg_iodc:clai/ ./msg_iodc:clm/ ./msg_iodc:cth/ ./msg_iodc:fire/ ./msg_iodc:ir039/ ./msg_iodc:ir108/ ./msg_iodc:mpe/ ./msg_iodc:rgb_airmass/ ./msg_iodc:rgb_ash/ ./msg_iodc:rgb_convection/ ./msg_iodc:rgb_dust/ ./msg_iodc:rgb_eview/ ./msg_iodc:rgb_fog/ ./msg_iodc:rgb_microphysics/ ./msg_iodc:rgb_natural/ ./msg_iodc:rgb_naturalenhncd/ ./msg_iodc:rgb_snow/ ./msg_iodc:rgb_tropicalairmass/ ./msg_iodc:vis006/ ./msg_iodc:wv062/ ./osmgray:builtup_area/ ./osmgray:land_polygons/ ./osmgray:ne_10m_admin_0_boundary_lines_land/ ./osmgray:ne_10m_admin_0_countries_points/ ./osmgray:ne_10m_admin_1_states_provinces_lines/ ./osmgray:ne_10m_bathymetry/ ./osmgray:osm_places/ ./osmgray:simplified_land_polygons/ ./osmgray:waterareas/ ./osmgray:waterways/ ./osmgray:dark/ ./osmgray:dark_bg/ ./osmgray:dark_labels/ ./osmgray:light/ ./osmgray:light_bg/ ./osmgray:light_labels/ ./copernicus:daily_sentinel3ab_olci_l1_rgb_fulres/ ./copernicus:daily_sentinel3ab_olci_l2_rgb_fulres/ ./copernicus:daily_sentinel3ab_olci_l2_rgb_fulres_histogram/ ./copernicus:daily_sentinel3ab_slstr_l2p_sst_fullres/ ./copernicus:daily_sentinel3ab_slstr_l2p_sst_fullres_histogram/ ./copernicus:daily_sentinel3ab_olci_l2_chl_fullres/ ./copernicus:daily_sentinel3ab_olci_l2_chl_fullres_histogram/ ./copernicus:sentinel3a_olci_l1_rgb_fullres/ ./copernicus:sentinel3a_olci_l2_chl_fullres/ ./copernicus:sentinel3a_olci_l2_chl_fullres_histogram/ ./copernicus:sentinel3b_olci_l1_rgb_fullres/ ./copernicus:sentinel3b_olci_l2_chl_fullres/ ./copernicus:sentinel3b_olci_l2_chl_fullres_II/ ./copernicus:sentinel3b_olci_l2_chl_fullres_histogram/ ./copernicus:sentinel3b_slstr_l2p_sst_fullres/ ./copernicus:sentinel3b_slstr_l2p_sst_fullres_histogram/ ./copernicus:sentinel3a_slstr_l2p_sst_fullres/ ./copernicus:sentinel3a_slstr_l2p_sst_fullres_histogram/ ./mumi:wideareacoverage_rgbnatcolour_jpeg/ ./mumi:wideareacoverage_rgbnatcolour/ ./mumi:worldcloudmap_ir108/
...
```
The directory in example is already empty from previous runs. What is the meaning of all the layer names printed? Can you make it explicit? To improve readability can you also print one layer per line instead of all layers names in a single line (maybe add a \t at the beginning of each line for the layer names so that is clear that the log belongs to that specific task)?

### For 2:
We're leaving behind all the empty directories. Over time, the list of layers will grow indefinetly. For instance in DEV we have directories with timestamp ` '#TIME#2019-11-27T23:45:00#000Z'`
Can we limit this? First thing that comes to my mind is:
- extract the max retention time from the layer.properties at the beginning of the script
- compare the timestamp of a directory against it. If the directory timestamp is "older than" the max retention for all layers then just remove the entire directory 

What do you think?

**Comment by jemacchi** [comment]
@randomorder 
**For 1:**
yes I can check how to improve logging (today it was listing all the layers since the rm command was executed as an unique command for all the layers that share the same retention time. 

**For 2:**
It's possible do it in that way, but question: what if instead of that we add a script that removes all the time-dimension folders which have no content inside ? (it's a simple think that it does not requires any parsing or calculation). Then we can add a second argument  -cronjob definition- in the mapcache-cleancache container in the STS (so, we can run that script as many times as required, and no need to base the logic of that clean based on the time-dimension).
My concern here is: what if someone starts playing with the retention-time back and forward with values/times ? (the other way it's simple: time-dimension folder without any folder inside...remove it, and let mapcache recreate if required/asked)
Sounds too crazy ? (I'm thinking that way since the sidecar container allows us to run ANY cronjob -script- we define/include in the docker image - 1 or many -  for cleaning or it could be for logging or any task we want, logging metrics, periodical validations, check status, anything)

**Comment by randomorder** [comment]
> **For 2:**
> It's possible do it in that way, but question: what if instead of that we add a script that removes all the time-dimension folders which have no content inside ? (it's a simple think that it does not requires any parsing or calculation).

Sounds good to me

**Comment by jemacchi** [comment]
@randomorder 
I have updated script in a way that it's easier to read the log, and at the same time we can clean time-dimension folders removed while cleaning layer folders cache (note: it's in middle of what you suggested and the parallel script running for cleaning those time-dimensions folders. Basically, if the time-dimension folder is empty after deleting layers-folders, then we try to remove the directory using -d parameter, if success, let us know, if it's not empty then that implies that a different "retention_key" value is still valid)

[PullRequest](https://github.com/geosolutions-it/eumetsat-EOWS/pull/465)

This does not invalidate the idea of an extra script for time-dimension folders removal (but it's not necessary to create it. I'm keeping the idea here since it could be usable for another kind of stuff)

**Comment by randomorder** [comment]
merged. Please let me know when ready to test in k8s DEV

**Comment by jemacchi** [comment]
@randomorder it's on DEV env (used sae tag for the docker image, so no need to update helm charts. Image was pushed, and scripts tested)

**Comment by simboss** [comment]
@randomorder one last modification has been asked to have a default global retention value that can be overridden by layer specific retention values.

Let's estimate this before implementing it.

**Comment by randomorder** [comment]
As discussed in our call @agpenton 
please move on with the changes to the script according to[ the tech note shared with EUM](https://docs.google.com/document/d/14yUBxzN0AtRVhfBwWwAlA4QxwtYNgS3VqDUUP2dBWYw/edit)
If you have any questions let me know

**Comment by randomorder** [comment]
keep us posted @agpenton !

**Comment by agpenton** [comment]
@randomorder about cleaning the old tiles from PostgreSQL and the storage, which one you consider is the best approach?
- [ ] query the path and create a list with all entries to delete, then delete the files after the entries in the database.
- [ ] query the path and create a list with all entries to delete, then delete the entries in the database after the files.
- [ ] query the path, then delete each file and database entries on the fly.

**Comment by simboss** [comment]
Always delete files after you deleted the rows in the DBMS index.
If you do the opposite I might still be able to query the files from GeoServer (as they are referenced in the index) and I would get an error afterwards.

Notice also that deleting files might fail due to files being locked, hence we might need to retry a few times and then eventually log that they weren't deleted, in the unfortunate case that this happens.

**Comment by randomorder** [comment]
remeber to log status updates in the issue please

**Comment by agpenton** [comment]
@randomorder I've pushed the script for the cleaning of the imageMosaic is here https://github.com/geosolutions-it/eumetsat-EOWS/tree/imageMosaic-clean-script.

**Comment by agpenton** [comment]
@randomorder the PR for the code https://github.com/geosolutions-it/eumetsat-EOWS/pull/477

**Comment by randomorder** [comment]
Please check my comments to the PR @agpenton

**Comment by agpenton** [comment]
@randomorder the PR is ready to review https://github.com/geosolutions-it/eumetsat-EOWS/pull/477
